chia/cmds/cmd_classes.py

```diff
@@ -0,0 +1,322 @@
+from __future__ import annotations
+
+import asyncio
+import collections
+import inspect
+import sys
+from contextlib import asynccontextmanager
+from dataclasses import MISSING, dataclass, field, fields
+from typing import (
+    Any,
+    AsyncIterator,
+    Callable,
+    Dict,
+    List,
+    Optional,
+    Protocol,
+    Type,
+    Union,
+    get_args,
+    get_origin,
+    get_type_hints,
+)
+
+import click
+from typing_extensions import dataclass_transform
+
+from chia.cmds.cmds_util import get_wallet_client
+from chia.rpc.wallet_rpc_client import WalletRpcClient
+from chia.types.blockchain_format.sized_bytes import bytes32
+from chia.util.byte_types import hexstr_to_bytes
+from chia.util.streamable import is_type_SpecificOptional
+
+SyncCmd = Callable[..., None]
+
+
+class SyncChiaCommand(Protocol):
+    def run(self) -> None: ...
+
+
+class AsyncChiaCommand(Protocol):
+    async def run(self) -> None: ...
+
+
+ChiaCommand = Union[SyncChiaCommand, AsyncChiaCommand]
+
+
+def option(*param_decls: str, **kwargs: Any) -> Any:
+    if sys.version_info < (3, 10):  # versions < 3.10 don't know about kw_only and they complain about lacks of defaults
+        # Can't get coverage on this because we only test on one version
+        default_default = None  # pragma: no cover
+    else:
+        default_default = MISSING
+
+    return field(  # pylint: disable=invalid-field-call
+        metadata=dict(
+            option_args=dict(
+                param_decls=tuple(param_decls),
+                **kwargs,
+            ),
```

Comment:
I'm not necessarily opposed, but to me it strikes me as an extra layer of complication. We parse stuff that comes in as a kwargs dict into a class and then later, when passing that stuff to click, we have to reserialize it into a dict.  Most of it remains untyped the whole time anyways.

Your comment on line 210 https://github.com/Chia-Network/chia-blockchain/pull/17851#discussion_r1567767436 makes me think that I have maybe done a poor job of writing clear code here because the idea is that every kwarg that gets passed to the `option` helper is supposed to also be passed to `click.option` with the exception of a few that we assign new special meaning to.

Reply:
> makes me think that I have maybe done a poor job of writing clear code

Naw, I got that most of it is passthrough and I expect that if a class is used that one of the attributes would be `Dict[str, Any]` to pass through to click.  I would just prefer that form with hinting for what little we can instead of this form with special cased unhinted dict entries.  I would expect the class to have well defined attributes for values that are not passed through to click at all and probably also for entries that we intercept on the way to click (`multiple`, others?).

Reply:
I took a crack at implementing this: dc8b77bc5d but also included a commit to revert it because I wasn't personally a fan. It felt a bit messier to me and there was also a quirk where we actually distinguish between not specifying a default and specifying `None` which a class is ill equipped for so we actually had to mostly call `option.kwargs["default"]` in most places anyways. The benefit was typing on two bools as far as I could tell. I think just from a methodological(?) standpoint the dict makes sense here because it is a layer that mostly examines values rather modifies them and pushing stuff into a class just seems to be asking for trouble rather than just (mostly) forwarding the dict to the `click.option` layer.